### PR TITLE
Add Delaware Itemized Deduction

### PR DESCRIPTION
**Pull request by Kiki-67**
### Summary
📝📄🧮

This pull request adds support for Delaware itemized deductions in the policy engine. It creates a new parameter file for the charity deduction, and modifies the formula for the `de_itemized_deductions` variable to use the new parameter and follow the state rules.

> _`de_itemized_deductions`_
> _New formula and file_
> _Autumn leaves fall gently_

### Walkthrough
*  Update changelog entry to reflect new feature ([link](https://github.com/PolicyEngine/policyengine-us/pull/3223/files?diff=unified&w=0#diff-5885042762833a4aef51405a4036067e4e6aefa7a5091a2987d1d9d1fc2647e8R1-R4))
*  Add parameter file for Delaware deduction for charity ([link](https://github.com/PolicyEngine/policyengine-us/pull/3223/files?diff=unified&w=0#diff-f6aebdb47d07a8ef449a3bbca11f0951afdbba074f153cd98df7e85af402bf9dR1-R13))
*  Modify formula for `de_itemized_deductions` variable to use new parameter and exclude federal deductions for charity and interest ([link](https://github.com/PolicyEngine/policyengine-us/pull/3223/files?diff=unified&w=0#diff-0c04c8fa14e5da4bc5ecc058e3637aefd4d2449da90cc5a3e7600e9b0d2c1dc0L17-R44))

### CODE REVIEW

File: policyengine_us/parameters/gov/states/de/tax/income/subtractions/exclusions/itemized/charity.yaml [hunk 1]
```diff
@@ -0,0 +1,13 @@
+description: Delaware caps the itemized deductions for charity at this amount.
+values:
+  2021-01-01: 250
+metadata:
+  label: Delaware deduction for charity
+  unit: currency-USD
+  reference:
+    - title: "Delaware Resident Schedule A - Itemized Deductions for 2022"
+      href: https://revenuefiles.delaware.gov/2022/TY22_PIT-RSA_2022-02_PaperInteractive.pdf
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/de/tax/income/subtractions/exclusions/itemized/charity.yaml:9]: need pdf page

File: policyengine_us/parameters/gov/states/de/tax/income/subtractions/exclusions/itemized/charity.yaml [hunk 2]
```diff
@@ -0,0 +1,13 @@
+description: Delaware caps the itemized deductions for charity at this amount.
+values:
+  2021-01-01: 250
+metadata:
+  label: Delaware deduction for charity
+  unit: currency-USD
+  reference:
+    - title: "Delaware Resident Schedule A - Itemized Deductions for 2022"
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/de/tax/income/subtractions/exclusions/itemized/charity.yaml:8]:
```suggestion
    - title: Delaware Resident Schedule A - Itemized Deductions for 2022
```

File: policyengine_us/parameters/gov/states/de/tax/income/subtractions/exclusions/itemized/charity.yaml [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+      href: https://revenuefiles.delaware.gov/2022/TY22_PIT-RSA_2022-02_PaperInteractive.pdf
+    - title: "Delaware Resident Schedule A - Itemized Deductions for 2021"
+      href: https://revenuefiles.delaware.gov/2021/TY21_PIT-RSA_2021-01_PaperInteractive.pdf
+    - title: Delaware State Taxes, Part II, Chapter 11, Subchapter 2, § 1109
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/de/tax/income/subtractions/exclusions/itemized/charity.yaml:12]: specific subparagraph of the law

Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/de/tax/income/subtractions/exclusions/itemized/charity.yaml:12]: double check references

File: policyengine_us/parameters/gov/states/de/tax/income/subtractions/exclusions/itemized/charity.yaml
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/de/tax/income/subtractions/exclusions/itemized/charity.yaml:1]: double check the parameter

File: changelog_entry.yaml [hunk 4]
```diff
@@ -0,0 +1,4 @@
+- bump: minor
+  changes:
+    added:
+    - Add Delaware itemized deductions.
```
Comment (PavelMakarchuk) [inline changelog_entry.yaml:4]:
```suggestion
    - Delaware itemized deductions.
```

File: policyengine_us/variables/gov/states/de/tax/income/deductions/itemized/de_itemized_deductions.py [hunk 5]
```diff
@@ -14,7 +14,31 @@ class de_itemized_deductions(Variable):
...
+    def formula(tax_unit, period, parameters):
+        p = parameters(
+            period
+        ).gov.states.de.tax.income.subtractions.exclusions.itemized
+        deductions = [
+            deduction
+            for deduction in p.itemized_deductions
+            if deduction
+            not in [
+                "interest_deduction",
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/de/tax/income/deductions/itemized/de_itemized_deductions.py:26]: exclude state and local taxes (salt_...)

File: policyengine_us/variables/gov/states/de/tax/income/deductions/itemized/de_itemized_deductions.py [hunk 6]
```diff
@@ -14,7 +14,31 @@ class de_itemized_deductions(Variable):
...
+        ]
+        federal_deductions = add(tax_unit, period, deductions)
+
+        de_donation = add(
+            tax_unit,
+            period,
+            ["charitable_cash_donations", "charitable_non_cash_donations"],
+        )
+
+        de_interest = add(
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/de/tax/income/deductions/itemized/de_itemized_deductions.py:38]:
```suggestion
        interest_deduction = add(
```

File: policyengine_us/variables/gov/states/de/tax/income/deductions/itemized/de_itemized_deductions.py [hunk 7]
```diff
@@ -14,7 +14,31 @@ class de_itemized_deductions(Variable):
...
+        p = parameters(
+            period
+        ).gov.states.de.tax.income.subtractions.exclusions.itemized
+        deductions = [
+            deduction
+            for deduction in p.itemized_deductions
+            if deduction
+            not in [
+                "interest_deduction",
+                "charitable_deduction",
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/de/tax/income/deductions/itemized/de_itemized_deductions.py:27]:
```suggestion
```

File: policyengine_us/variables/gov/states/de/tax/income/deductions/itemized/de_itemized_deductions.py [hunk 8]
```diff
@@ -14,7 +14,31 @@ class de_itemized_deductions(Variable):
...
+            ]
+        ]
+        federal_deductions = add(tax_unit, period, deductions)
+
+        de_donation = add(
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/de/tax/income/deductions/itemized/de_itemized_deductions.py:32]: dont need this

File: policyengine_us/variables/gov/states/de/tax/income/deductions/itemized/de_itemized_deductions.py
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/de/tax/income/deductions/itemized/de_itemized_deductions.py:1]: need unit test

File: policyengine_us/parameters/gov/states/de/tax/income/subtractions/exclusions/itemized/charity.yaml
Comment (PavelMakarchuk) [reply policyengine_us/parameters/gov/states/de/tax/income/subtractions/exclusions/itemized/charity.yaml:1]: probably delete

File: policyengine_us/variables/gov/states/de/tax/income/deductions/itemized/de_itemized_deductions.py
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/de/tax/income/deductions/itemized/de_itemized_deductions.py:1]: we have a circular reference in `de_income_tax` the variables refer to each other in the computation, please take a look at the error message try to resolve

File: policyengine_us/variables/gov/states/de/tax/income/deductions/itemized/de_itemized_deductions.py [hunk 9]
```diff
@@ -7,14 +7,35 @@ class de_itemized_deductions(Variable):
...
+        filing_status = tax_unit("filing_status", period)
+        deductions = [
+            deduction
+            for deduction in p.itemized_deductions
+            if deduction not in ["interest_deduction", "salt_deduction"]
+        ]
+        federal_deductions = add(tax_unit, period, deductions)
+
+        salt_sales_or_income = tax_unit(
+            "state_and_local_sales_or_income_tax", period
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/de/tax/income/deductions/itemized/de_itemized_deductions.py:24]:
```suggestion
```

File: policyengine_us/variables/gov/states/de/tax/income/deductions/itemized/de_itemized_deductions.py [hunk 10]
```diff
@@ -7,14 +7,35 @@ class de_itemized_deductions(Variable):
...
+        )
+
+        total_salt = (
+            add(tax_unit, period, ["real_estate_taxes"]) + salt_sales_or_income
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/de/tax/income/deductions/itemized/de_itemized_deductions.py:28]:
```suggestion
            add(tax_unit, period, ["real_estate_taxes"])
```

File: policyengine_us/variables/gov/states/de/tax/income/deductions/itemized/de_itemized_deductions.py [hunk 11]
```diff
@@ -7,14 +7,35 @@ class de_itemized_deductions(Variable):
...
+            for deduction in p.itemized_deductions
+            if deduction not in ["interest_deduction", "salt_deduction"]
+        ]
+        federal_deductions = add(tax_unit, period, deductions)
+
+        salt_sales_or_income = tax_unit(
+            "state_and_local_sales_or_income_tax", period
+        )
+
+        total_salt = (
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/de/tax/income/deductions/itemized/de_itemized_deductions.py:27]:
```suggestion
        real_estate_tax = (
```

File: policyengine_us/variables/gov/states/de/tax/income/deductions/itemized/de_itemized_deductions.py [hunk 12]
```diff
@@ -7,14 +7,35 @@ class de_itemized_deductions(Variable):
     label = "Delaware itemized deductions"
     unit = USD
     definition_period = YEAR
-    reference = (
-        "https://revenuefiles.delaware.gov/2022/PIT-RES_TY22_2022-02_Instructions.pdf#page=7"
-        "https://revenuefiles.delaware.gov/2022/PIT-RES_TY22_2022-02_Instructions.pdf#page=11"
-        "https://casetext.com/statute/delaware-code/title-30-state-taxes/part-ii-income-inheritance-and-estate-taxes/chapter-11-personal-income-tax/subchapter-ii-resident-individuals/section-1109-itemized-deductions-for-application-of-this-section-see-66-del-laws-c-86-section-8"
-    )
+    reference = "https://revenuefiles.delaware.gov/2022/TY22_PIT-RSA_2022-02_PaperInteractive.pdf"
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/de/tax/income/deductions/itemized/de_itemized_deductions.py:10]: add the legal as a second reference

Comment (PavelMakarchuk) [reply policyengine_us/variables/gov/states/de/tax/income/deductions/itemized/de_itemized_deductions.py:10]: https://casetext.com/statute/delaware-code/title-30-state-taxes/part-ii-income-inheritance-and-estate-taxes/chapter-11-personal-income-tax/subchapter-ii-resident-individuals/section-1109-itemized-deductions

File: policyengine_us/variables/gov/states/de/tax/income/deductions/itemized/de_itemized_deductions.py [hunk 13]
```diff
@@ -8,13 +8,31 @@ class de_itemized_deductions(Variable):
...
+        deductions = [
+            deduction
+            for deduction in p.itemized_deductions
+            if deduction not in ["interest_deduction", "salt_deduction"]
+        ]
+        federal_deductions = add(tax_unit, period, deductions)
+
+        real_estate_tax = add(tax_unit, period, ["real_estate_taxes"])
+
+        salt_amount = min_(
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/de/tax/income/deductions/itemized/de_itemized_deductions.py:28]:
```suggestion
        capped_ real_estate_tax = min_(
```

File: policyengine_us/variables/gov/states/de/tax/income/deductions/itemized/de_itemized_deductions.py [hunk 14]
```diff
@@ -8,13 +8,31 @@ class de_itemized_deductions(Variable):
...
+            real_estate_tax, p.itemized.salt_and_real_estate.cap[filing_status]
+        )
+
+        interest = add(
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/de/tax/income/deductions/itemized/de_itemized_deductions.py:32]:
```suggestion
        interest_deduction = add(
```

**Review by PavelMakarchuk — APPROVED** [review]

File: policyengine_us/tests/policy/baseline/gov/states/de/tax/income/subtractions/de_itemized_deductions.yaml [hunk 15]
```diff
@@ -0,0 +1,57 @@
+- name: filling status = SEPERATE, total salt < salt cap
```
Comment (MaxGhenis) [inline policyengine_us/tests/policy/baseline/gov/states/de/tax/income/subtractions/de_itemized_deductions.yaml:1]: and other test names
```suggestion
- name: Separate filing status, total SALT < SALT cap
```

File: policyengine_us/variables/gov/states/de/tax/income/deductions/itemized/de_itemized_deductions.py [hunk 16]
```diff
@@ -8,13 +8,31 @@ class de_itemized_deductions(Variable):
     unit = USD
     definition_period = YEAR
     reference = (
-        "https://revenuefiles.delaware.gov/2022/PIT-RES_TY22_2022-02_Instructions.pdf#page=7"
-        "https://revenuefiles.delaware.gov/2022/PIT-RES_TY22_2022-02_Instructions.pdf#page=11"
-        "https://casetext.com/statute/delaware-code/title-30-state-taxes/part-ii-income-inheritance-and-estate-taxes/chapter-11-personal-income-tax/subchapter-ii-resident-individuals/section-1109-itemized-deductions-for-application-of-this-section-see-66-del-laws-c-86-section-8"
+        "https://revenuefiles.delaware.gov/2022/TY22_PIT-RSA_2022-02_PaperInteractive.pdf",
```
Comment (MaxGhenis) [inline policyengine_us/variables/gov/states/de/tax/income/deductions/itemized/de_itemized_deductions.py:11]:
```suggestion
        "https://revenuefiles.delaware.gov/2022/TY22_PIT-RSA_2022-02_PaperInteractive.pdf",  # § 1109
```

File: policyengine_us/variables/gov/states/de/tax/income/deductions/itemized/de_itemized_deductions.py [hunk 17]
```diff
@@ -8,13 +8,31 @@ class de_itemized_deductions(Variable):
...
-    # Per Law: Self employed filers can deduct the health insurance premiums
-    # less the amount allowed as a deduction on the federal return
-    # We omit this because it is not on the tax form.
+    def formula(tax_unit, period, parameters):
+        p = parameters(period).gov.irs.deductions
+        filing_status = tax_unit("filing_status", period)
+        deductions = [
+            deduction
+            for deduction in p.itemized_deductions
+            if deduction not in ["interest_deduction", "salt_deduction"]
```
Comment (MaxGhenis) [inline policyengine_us/variables/gov/states/de/tax/income/deductions/itemized/de_itemized_deductions.py:22]: why subtract and re-add the interest deduction?

please leave comments in each code section pointing to the legal code subsections and tax form lines

**Review by MaxGhenis — CHANGES_REQUESTED** [review]

File: policyengine_us/variables/gov/states/de/tax/income/deductions/itemized/de_itemized_deductions.py (hunk 17 above)
Comment (PavelMakarchuk) [reply policyengine_us/variables/gov/states/de/tax/income/deductions/itemized/de_itemized_deductions.py:22]: assumed that `mortgage_interest` and `investment_income_form_4952` exceed the scope of `interest_deduction`, but looking at it, will change back

Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/de/tax/income/deductions/itemized/de_itemized_deductions.py:22]:
```suggestion
            if deduction not in ["salt_deduction"]
```

File: policyengine_us/variables/gov/states/de/tax/income/deductions/itemized/de_itemized_deductions.py [hunk 18]
```diff
@@ -8,13 +8,31 @@ class de_itemized_deductions(Variable):
...
+        ]
+        federal_deductions = add(tax_unit, period, deductions)
+
+        real_estate_tax = add(tax_unit, period, ["real_estate_taxes"])
+
+        capped_real_estate_tax = min_(
+            real_estate_tax, p.itemized.salt_and_real_estate.cap[filing_status]
+        )
+
+        interest_deduction = add(
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/de/tax/income/deductions/itemized/de_itemized_deductions.py:32]: delete this part

File: policyengine_us/variables/gov/states/de/tax/income/deductions/itemized/de_itemized_deductions.py [hunk 19]
```diff
@@ -8,13 +8,31 @@ class de_itemized_deductions(Variable):
...
+            tax_unit,
+            period,
+            ["mortgage_interest", "investment_income_form_4952"],
+        )
+
+        return federal_deductions + capped_real_estate_tax + interest_deduction
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/de/tax/income/deductions/itemized/de_itemized_deductions.py:38]:
```suggestion
        return federal_deductions + capped_real_estate_tax
```

**Pull request merged**